src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -542,35 +585,56 @@ private ValueTask<FlushResult> WriteDataAndTrailersAsync(Http2Stream stream, in
         }
     }
 
-    private void FinishWritingHeaders(int streamId, int payloadLength, bool done)
+    private void SplitHeaderFramesToOutput(int streamId, HPackHeaderWriter.HeaderWriteResult done, bool isFramePrepared)
     {
-        var buffer = _headerEncodingBuffer.AsSpan();
-        _outgoingFrame.PayloadLength = payloadLength;
-        if (done)
-        {
-            _outgoingFrame.HeadersFlags |= Http2HeadersFrameFlags.END_HEADERS;
-        }
-
-        WriteHeaderUnsynchronized();
-        _outputWriter.Write(buffer.Slice(0, payloadLength));
-
-        while (!done)
+        var dataToFrame = _headerEncodingBuffer.WrittenSpan;
+        var shouldPrepareFrame = !isFramePrepared;
+        while (dataToFrame.Length > 0)
         {
-            _outgoingFrame.PrepareContinuation(Http2ContinuationFrameFlags.NONE, streamId);
-
-            done = HPackHeaderWriter.ContinueEncodeHeaders(_hpackEncoder, _headersEnumerator, buffer, out payloadLength);
-            _outgoingFrame.PayloadLength = payloadLength;
+            if (shouldPrepareFrame)
+            {
+                _outgoingFrame.PrepareContinuation(Http2ContinuationFrameFlags.NONE, streamId);
+            }
+            else
+            {
+                shouldPrepareFrame = true;
```

Comment:
Yes, I wanted to get rid of this. but I did not find a good way to do so. Both the initial call to `WriteResponseHeadersUnsynchronized` and `WriteDataAndTrailersAsync` calls `_outgoingFrame.PrepareHeaders`, so if you end up splitting the initial frame (this can only happen for the trailers in practice), as you write every subsequent frames we need to *Prepare* the continuation. I am happy to set it unconditionally, but cannot get rid of the branching itself (I only had it set conditionally because the branching was already there).